runtime/tsan.c

```diff
@@ -0,0 +1,240 @@
+/*                                 OCaml                                  */
+/*                                                                        */
+/*               Fabrice Buoro and Olivier Nicole, Tarides                */
+/*                                                                        */
+/*   Copyright 2023 Tarides                                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+
+#ifdef NATIVE_CODE
+
+#ifndef DEBUG
+#undef TSAN_DEBUG
+#endif
+
+#define UNW_LOCAL_ONLY
+#include <libunwind.h>
+
+#include "caml/tsan.h"
+#include "caml/mlvalues.h"
+#include "caml/misc.h"
+#include "caml/frame_descriptors.h"
+#include "caml/fiber.h"
+#include "caml/domain_state.h"
+#include "caml/stack.h"
+#include "caml/config.h"
+#ifdef TSAN_DEBUG
+#include <execinfo.h> /* For backtrace_symbols */
+#endif
+
+extern void __tsan_func_exit(void*);
+#if defined(__GNUC__) && !defined(__clang__)
+/* GCC already has __tsan_func_entry declared for some reason */
+#else
+extern void __tsan_func_entry(void*);
+#endif
+
+Caml_inline void caml_tsan_debug_log_pc(const char* msg, uintnat pc)
+{
+#ifdef TSAN_DEBUG
+    char **sym_names = backtrace_symbols((void **)&pc, 1);
+    fprintf(stderr, "%s %s\n", msg, sym_names[0]);
+    free(sym_names);
+#else
+    (void)msg; (void)pc;
+#endif
+}
+
+void caml_tsan_exn_func_exit(uintnat pc, char* sp, char* trapsp)
+{
+  caml_domain_state* domain_state = Caml_state;
+  caml_frame_descrs fds = caml_get_frame_descrs();
+  uintnat next_pc = pc;
+
+  /* iterate on each frame  */
+  while (1) {
+    frame_descr* descr = caml_next_frame_descriptor(fds, &next_pc, &sp,
+        domain_state->current_stack);
+
+    if (descr == NULL) {
+      return;
+    }
+
+    /* Stop when we reach the current exception handler */
+    if (sp > trapsp) {
+      break;
+    }
+
+    caml_tsan_debug_log_pc("forced__tsan_func_exit for", pc);
+    __tsan_func_exit(NULL);
+    pc = next_pc;
+  }
+}
+
+void caml_tsan_exn_func_exit_c(char* limit)
+{
+  unw_context_t uc;
+  unw_cursor_t cursor;
+  unw_word_t sp;
+#ifdef TSAN_DEBUG
+  unw_word_t prev_pc;
+#endif
+  int ret;
+
+  ret = unw_getcontext(&uc);
+  if (ret != 0)
+    caml_fatal_error("unw_getcontext failed with code %d", ret);
+  ret = unw_init_local(&cursor, &uc);
+  if (ret != 0)
+    caml_fatal_error("unw_init_local failed with code %d", ret);
+
+  while (1) {
+#ifdef TSAN_DEBUG
+    if (unw_get_reg(&cursor, UNW_REG_IP, &prev_pc) < 0) {
+      caml_fatal_error("unw_get_reg failed with code %d", ret);
```

Comment:
just a minor thing: add in IP in this error string (and SP in the one a few lines below) so you can distinguish the source of the log messages at runtime in the future.

Reply:
Done, thanks.